gc/gc-base/src/main/java/org/projectnessie/gc/base/GCImpl.java

```diff
@@ -0,0 +1,189 @@
+   * for it. This is an optimization to avoid traversing all the commits.
+   *
+   * <p>Collect bloom filter per content id from each task and merge them.
+   *
+   * <h2>Algorithm for identifying the expired contents and return the list of globally expired
+   * contents per content id per reference </h2>
+   *
+   * <p>Walk through each reference(both live and dead) distributively (one spark task for each
+   * reference).
+   *
+   * <p>For each commit in the reference (use DETACHED reference to fetch commits from dead
+   * reference) check it against bloom filter to decide whether its contents in put operation are
+   * globally expired or not. If globally expired, Add the contents to the expired output for this
+   * content id for this reference.
+   *
+   * <p>Overall the contents after or equal to cutoff time and the contents that are mapped to
+   * commit head of live keys at the time of cutoff timestamp will be retained.
+   *
+   * @param session spark session for distributed computation
+   * @return {@link IdentifiedResult} object having expired contents per content id.
+   */
+  public IdentifiedResult identifyExpiredContents(SparkSession session) {
+    try (NessieApiV1 api = GCUtil.getApi(gcParams.getNessieClientConfigs())) {
+      DistributedIdentifyContents distributedIdentifyContents =
+          new DistributedIdentifyContents(session, gcParams);
+      List<Reference> liveReferences = api.getAllReferences().get().getReferences();
+      Map<Reference, Instant> droppedReferenceTimeMap = collectDeadReferences(api);
+      List<Reference> allRefs = new ArrayList<>(liveReferences);
+      if (droppedReferenceTimeMap.size() > 0) {
+        allRefs.addAll(droppedReferenceTimeMap.keySet());
+      }
+      long bloomFilterSize =
+          gcParams.getBloomFilterExpectedEntries() == null
+              ? getTotalCommitsInDefaultReference(api)
+              : gcParams.getBloomFilterExpectedEntries();
+      // Identify the live contents and return the bloom filter per content-id
+      Map<String, ContentBloomFilter> liveContentsBloomFilterMap =
+          distributedIdentifyContents.getLiveContentsBloomFilters(
+              allRefs, bloomFilterSize, droppedReferenceTimeMap);
+      // Identify the expired contents
+      return distributedIdentifyContents.getIdentifiedResults(liveContentsBloomFilterMap, allRefs);
+    }
+  }
+
+  private long getTotalCommitsInDefaultReference(NessieApiV1 api) {
+    ReferenceMetadata defaultRefMetadata = null;
+    try {
+      defaultRefMetadata =
+          api.getReference()
+              .refName(api.getDefaultBranch().getName())
+              .fetch(FetchOption.ALL)
+              .get()
+              .getMetadata();
+    } catch (NessieNotFoundException ex) {
+      throw new RuntimeException(ex);
+    }
+    return (defaultRefMetadata != null && defaultRefMetadata.getNumTotalCommits() != null)
```

Comment:
when the fetch option is not mentioned it is definitely null.
But as it is a @Nullable in immutables declaration, IDE thinks it can be null in my case also.
To avoid warnings added a null check.

Reply:
`defaultRefMetadata` should always be non-null here - everything else is an error.

Reply:
Removed null checks. But IDE errors will confuse the readers. They think it can be null.